pallets/settlement/src/lib.rs

```diff
@@ -971,6 +1071,97 @@ impl<T: Trait> Module<T> {
         Ok(instruction_counter)
     }
 
+    pub fn base_confidential_add_instruction(
+        did: IdentityId,
+        venue_id: u64,
+        settlement_type: SettlementType<T::BlockNumber>,
+        valid_from: Option<T::Moment>,
+        legs: Vec<Leg<T::Balance>>,
+    ) -> Result<u64, DispatchError> {
+        // Check whether the no. of legs within the limit or not.
+        ensure!(
+            u32::try_from(legs.len()).unwrap_or_default() <= T::MaxLegsInAInstruction::get(),
+            Error::<T>::LegsCountExceededMaxLimit
+        );
+        // Check if a venue exists and the sender is the creator of the venue
+        let mut venue = Self::venue_info(venue_id);
+        ensure!(venue.creator == did, Error::<T>::Unauthorized);
+
+        // Prepare data to store in storage
+        // NB Instruction counter starts from 1
+        let instruction_counter = Self::instruction_counter();
+        let mut counter_parties = BTreeSet::new();
+        let mut tickers = BTreeSet::new();
+        // This is done to create a list of unique CP and tickers involved in the instruction.
+        for leg in &legs {
+            ensure!(leg.from != leg.to, Error::<T>::SameSenderReceiver);
+            counter_parties.insert(leg.from);
+            counter_parties.insert(leg.to);
+            tickers.insert(leg.asset);
+        }
+
+        // Check if the venue has required permissions from token owners
+        for ticker in &tickers {
+            if Self::venue_filtering(ticker) {
+                ensure!(
+                    Self::venue_allow_list(ticker, venue_id),
+                    Error::<T>::UnauthorizedVenue
+                );
+            }
+        }
+
+        let instruction = Instruction {
+            instruction_id: instruction_counter,
+            venue_id: venue_id,
+            status: InstructionStatus::Pending,
+            settlement_type: settlement_type,
+            created_at: Some(<pallet_timestamp::Module<T>>::get()),
+            valid_from: valid_from,
+        };
+
+        // write data to storage
+        for counter_party in &counter_parties {
+            <UserAuths>::insert(
+                counter_party,
+                instruction_counter,
+                AuthorizationStatus::Pending,
+                // TODO: Store the account id here.
+            );
+        }
+
+        for (i, leg) in legs.iter().enumerate() {
+            <InstructionLegs<T>>::insert(
+                instruction_counter,
+                u64::try_from(i).unwrap_or_default(),
+                leg.clone(),
+            );
+        }
```

Comment:
I am not sure what is happening here. why zip `0u64`?

Reply:
Instead of getting a `usize` and converting it to a `u64` via `u64::try_from(i).unwrap_or_default()`, this produces `u64`s directly starting from `0`. So this is a simpler, and more direct approach.